File: python_pipeline/gar/exceedance.py
"""Exceedance correlations (Ang & Chen 2002) with the Hong-Tu-Zhou (2003)
asymmetry test. Replaces exceedence_correl.m (fixed), corrcoef12.m,
newey_west.m."""

from __future__ import annotations

import numpy as np
import pandas as pd
from scipy.stats import chi2


def newey_west_cov(Xi: np.ndarray, lags: int | None = None) -> np.ndarray:
    Tn = Xi.shape[0]
    if lags is None:
        lags = int(np.floor(4 * (Tn / 100.0) ** (2.0 / 9.0)))
    Xi = Xi - Xi.mean(axis=0)
    om = Xi.T @ Xi / Tn
    for l in range(1, lags + 1):
        w = 1.0 - l / (lags + 1.0)
        G = Xi[l:].T @ Xi[:-l] / Tn
        om += w * (G + G.T)
    return om


def exceedance_correl(X, Y, qc=None, min_n: int = 3) -> dict:
    """Lower tail: corr(X, Y | X <= Qx(1-q), Y <= Qy(1-q));
    Upper tail: corr(X, Y | X >= Qx(q),  Y >= Qy(q)),  q in [0.5, 0.95].

    Returns {"table": DataFrame(threshold, lower, upper, n_lower, n_upper),
             "teststat", "pval"}.
    """
    if qc is None:
        qc = np.arange(0.5, 0.951, 0.05)
    X = np.asarray(X, float)
    Y = np.asarray(Y, float)
    ok = np.isfinite(X) & np.isfinite(Y)
    X, Y = X[ok], Y[ok]
    Tn = len(X)

    rows, Xi_cols, rhodiff = [], [], []
    for q in qc:
        lo = (X <= np.quantile(X, 1 - q)) & (Y <= np.quantile(Y, 1 - q))
        up = (X >= np.quantile(X, q)) & (Y >= np.quantile(Y, q))
        n_lo, n_up = int(lo.sum()), int(up.sum())
        r_lo = r_up = np.nan
        if n_lo >= min_n and n_up >= min_n:
            r_lo = float(np.corrcoef(X[lo], Y[lo])[0, 1])
            r_up = float(np.corrcoef(X[up], Y[up])[0, 1])
            zx_up = (X - X[up].mean()) / X[up].std(ddof=1)
            zy_up = (Y - Y[up].mean()) / Y[up].std(ddof=1)
            zx_lo = (X - X[lo].mean()) / X[lo].std(ddof=1)
            zy_lo = (Y - Y[lo].mean()) / Y[lo].std(ddof=1)
            xi = (Tn / n_up * (zx_up * zy_up - r_up) * up
                  - Tn / n_lo * (zx_lo * zy_lo - r_lo) * lo)
            Xi_cols.append(xi)
            rhodiff.append(r_up - r_lo)
        rows.append({"threshold": round(float(q), 4), "lower": r_lo,
                     "upper": r_up, "n_lower": n_lo, "n_upper": n_up})

    teststat = pval = np.nan
    if rhodiff:
        Om = newey_west_cov(np.column_stack(Xi_cols))
        rd = np.asarray(rhodiff)
        try:
            teststat = float(Tn * rd @ np.linalg.solve(Om, rd))
            pval = float(chi2.sf(teststat, df=len(rd)))
        except np.linalg.LinAlgError:
            pass

    return {"table": pd.DataFrame(rows), "teststat": teststat, "pval": pval}

File: r_pipeline/R/exceedance.R
# exceedance.R -- exceedance correlations (Ang & Chen 2002) with the
# Hong-Tu-Zhou (2003) asymmetry test.
# Replaces: exceedence_correl.m (fixed version), corrcoef12.m, newey_west.m,
# plot_exceedence_correlation.m.

#' Newey-West long-run covariance of a T x k matrix of (demeaned) scores.
newey_west_cov <- function(Xi, lags = NULL) {
  Tn <- nrow(Xi)
  if (is.null(lags)) lags <- floor(4 * (Tn / 100)^(2 / 9))
  Xi <- scale(Xi, center = TRUE, scale = FALSE)
  om <- crossprod(Xi) / Tn
  if (lags > 0) {
    for (l in seq_len(lags)) {
      w <- 1 - l / (lags + 1)
      G <- crossprod(Xi[(l + 1):Tn, , drop = FALSE], Xi[1:(Tn - l), , drop = FALSE]) / Tn
      om <- om + w * (G + t(G))
    }
  }
  om
}

#' Exceedance correlations across quantile thresholds, with tail sample
#' sizes and the HTZ symmetry test.
#'
#' Lower tail: corr(X, Y | X <= Qx(1-q), Y <= Qy(1-q))
#' Upper tail: corr(X, Y | X >= Qx(q),   Y >= Qy(q)),  q in [0.5, 0.95]
#'
#' @param X,Y numeric vectors
#' @param qc  vector of upper-half thresholds (>= 0.5)
#' @param min_n minimum tail observations to compute a correlation
#' @return list(table = data.frame(threshold, lower, upper, n_lower, n_upper),
#'              teststat, pval)
exceedance_correl <- function(X, Y, qc = seq(0.5, 0.95, by = 0.05), min_n = 3L) {
  ok <- stats::complete.cases(X, Y)
  X <- X[ok]; Y <- Y[ok]
  Tn <- length(X)

  res <- data.frame(threshold = qc, lower = NA_real_, upper = NA_real_,
                    n_lower = NA_integer_, n_upper = NA_integer_)
  Xi <- NULL; rhodiff <- c(); used <- c()

  for (j in seq_along(qc)) {
    q <- qc[j]
    lo <- which(X <= stats::quantile(X, 1 - q) & Y <= stats::quantile(Y, 1 - q))
    up <- which(X >= stats::quantile(X, q)     & Y >= stats::quantile(Y, q))
    res$n_lower[j] <- length(lo); res$n_upper[j] <- length(up)
    if (length(lo) < min_n || length(up) < min_n) next

    r_lo <- stats::cor(X[lo], Y[lo]); r_up <- stats::cor(X[up], Y[up])
    res$lower[j] <- r_lo; res$upper[j] <- r_up

    # HTZ influence scores
    zx_up <- (X - mean(X[up])) / stats::sd(X[up])
    zy_up <- (Y - mean(Y[up])) / stats::sd(Y[up])
    zx_lo <- (X - mean(X[lo])) / stats::sd(X[lo])
    zy_lo <- (Y - mean(Y[lo])) / stats::sd(Y[lo])
    ind_up <- as.numeric(X >= stats::quantile(X, q)     & Y >= stats::quantile(Y, q))
    ind_lo <- as.numeric(X <= stats::quantile(X, 1 - q) & Y <= stats::quantile(Y, 1 - q))
    xi_j <- Tn / length(up) * (zx_up * zy_up - r_up) * ind_up -
            Tn / length(lo) * (zx_lo * zy_lo - r_lo) * ind_lo
    Xi <- cbind(Xi, xi_j)
    rhodiff <- c(rhodiff, r_up - r_lo)
    used <- c(used, q)
  }

  teststat <- NA_real_; pval <- NA_real_
  if (length(rhodiff) > 0) {
    om <- newey_west_cov(Xi)
    sol <- try(solve(om, rhodiff), silent = TRUE)
    if (!inherits(sol, "try-error")) {
      teststat <- Tn * sum(rhodiff * sol)
      pval <- 1 - stats::pchisq(teststat, df = length(rhodiff))
    }
  }
  list(table = res, teststat = teststat, pval = pval, thresholds_used = used)
}
